src/dac.rs

```diff
@@ -0,0 +1,248 @@
+//! Configure the internal DAC on the stm32f3xx.
+//! Incomplete, but includes basic operation.
+//!
+//! You should have the appropriate DAC pin set up as an analog input, to prevent
+//! parasitic power consumption. For example:
+//! ```rust
+//! let _dac1_pin = gpioa.pa4.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! let _dac2_pin = gpioa.pa5.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! ```
+
+use core::{convert::Infallible, fmt};
+
+use crate::{
+    gpio::{
+        gpioa::{PA4, PA5},
+        Analog,
+    },
+    pac,
+    rcc::APB1,
+};
+
+/// todo: DAC trait that will go in a PR to `embeddd-hal`.
+/// Once merged and released in that crate, remove from this module.
+/// [PR](https://github.com/rust-embedded/embedded-hal/pull/247)
+pub trait SingleChannelDac {
+    /// Error type returned by DAC methods
+    type Error;
+    type Word;
+
+    /// Output a constant signal, given a bit word.
+    fn try_set_value(&mut self, value: Self::Word) -> Result<(), Self::Error>;
+}
+
+/// This is an abstraction to ensure that the DAC output pin is configured
+/// as an analog output.
+pub trait Pins {}
+impl Pins for PA4<Analog> {}
+impl Pins for PA5<Analog> {}
+impl Pins for (PA4<Analog>, PA5<Analog>) {}
+
+#[derive(Clone, Copy, Debug)]
+/// Select the channel
+pub enum Channel {
+    /// Channel 1
+    One,
+    /// Channel 2
+    Two,
+}
+
+#[derive(Clone, Copy, Debug)]
+/// Three options are available to set DAC precision.
+pub enum DacBits {
+    /// Eight bit precision, right-aligned.
+    EightR,
+    /// 12-bit precision, left-aligned.
+    TwelveL,
+    /// 12-bit precision, right-aligned.
+    TwelveR,
+}
+
+#[derive(Clone, Copy)]
+/// Select a trigger, used by some features.
+pub enum Trigger {
+    /// Timer 6
+    Tim6,
+    /// Timers 3 or 8
+    Tim3_8,
+    /// Timer 7
+    Tim7,
+    /// Timer 15
+    Tim15,
+    /// Timger 2
+    Tim2,
+    /// Tim4 is Not available on DAC 2.
+    Tim4,
+    /// Eg, for interrupts
+    Exti9,
+    /// A software trigger
+    Swtrig,
+}
+
+impl Trigger {
+    pub fn bits(&self) -> u8 {
+        match self {
+            Self::Tim6 => 0b000,
+            Self::Tim3_8 => 0b001,
+            Self::Tim7 => 0b010,
+            Self::Tim15 => 0b011,
+            Self::Tim2 => 0b100,
+            Self::Tim4 => 0b101,
+            Self::Exti9 => 0b110,
+            Self::Swtrig => 0b111,
+        }
+    }
+}
+
+pub struct Dac {
+    regs: pac::DAC,
+    channel: Channel,
+    bits: DacBits,
+    vref: f32,
+}
+
+impl Dac {
+    /// Create a new DAC instance
+    pub fn new<P: Pins>(
+        regs: pac::DAC,
+        _pins: P,
+        channel: Channel,
```

Comment:
Now it is possible for the user to pass in the wrong pin for the given channel. Instead can we select the channel based on the pin? Could be done via trait methods on `Pins`, to allow you to do `P::enable_dac_channel()` and `P::disable_dac_channel()` with the correct channel.